webapp/src/domain/tests/Hooks.tsx

```diff
@@ -0,0 +1,204 @@
+import { TestDispatch } from './reducers';
+import { Hook } from "../hooks/reducers";
+import { TabFunctionsRef } from './Test'
+import {updateHooks} from './actions'
+import { ValueGetter } from '../../components/Editor/monaco/Editor'
+import {eventTypes} from "../hooks/AddHookModal";
+import * as api from "../hooks/api";
+
+function swap(array: any[], i1: number, i2: number) {
+    const temp = array[i1]
+    array[i1] = array[i2]
+    array[i2] = temp
+}
+
+type HookComponentFormProps = {
+    c: Hook,
+    onChange(): void,
+    isTester: boolean,
+}
+
+
+
+const HookComponentForm = ({ c, onChange, isTester } : HookComponentFormProps) => {
+    const [eventType,setEventType] = useState(c.type)
+
+    return (
+        <Form isHorizontal={true} style={{ gridGap: "2px", width: "100%", float: "left", marginBottom: "25px" }}>
+            <FormGroup label="Webhook URL" fieldId="url">
+                <TextInput value={ c.url || "" } placeholder="e.g. 'http://myserver/api/hook'"
+                           id="header"
+                           onChange={ value => {
+                               c.url = value
+                               onChange()
+                           }}
+                           validated={ !!c.url  && c.url.trim() !== "" ? "default" : "error" }
+                           isReadOnly={!isTester} />
+            </FormGroup>
+            <FormGroup label="Event Type" fieldId="event-type">
+                <FormSelect
+                    id="type"
+                    validated={"default"}
+                    value={eventType}
+                    onChange={ value => {
+                        c.type = value
+                        setEventType(value)
+                        onChange()
+                    }}
+                    aria-label="Event Type"
+                >
+                    {eventTypes.map((option, index)=>{
+                        return (<FormSelectOption
+                            key={index}
+                            value={option}
+                            label={option}  />)
+                    })}
+                </FormSelect>
+            </FormGroup>
+        </Form>
+    )
+
+}
+
+type HooksProps = {
+    testId: number,
+    // testWebHooks: TestWebHooks,
+    testOwner?: string,
+    funcsRef: TabFunctionsRef,
+    onModified(modified: boolean): void,
+}
+
+export default ({ testId, testOwner, funcsRef, onModified }: HooksProps) => {
+    const [testWebHooks, setTestWebHooks] = useState<Hook[]>([])
+    const isTester = useTester(testOwner)
+    const renderRefs = useRef(new Array<ValueGetter | undefined>(testWebHooks.length));
+
+    useEffect(() => {
+        if (!testId) {
+            return
+        }
+        api.fetchHooks(testId).then(
+            response => {
+                setTestWebHooks(response.map((h: Hook) => {
+                    let hd: Hook = {
+                        ...h,
+                        id: Number(h.id),
```

Comment:
idk if they are explicitly required, iirc the IDE was not resolving the correct type but that is probably an issue with the IDE rather than the type could not be inferred at runtime. I can look at removing the type cast

Reply:
`response` array is result of a call to `fetchApi()` which returns a `Promise<Any>`. so `response.map((h: Hook)` requires the typecast, unless we want to include return type as a param to `fetchApi()`.  That would be quite an intrusive change. Explicitly casting `let hd: Hook` can be dropped.